graf2d/primitives/v7/inc/ROOT/TDisplayItem.hxx

```diff
@@ -31,71 +31,43 @@ class TFrame;
 
 class TDisplayItem {
 protected:
-   std::string fObjectID;
-   std::string fOption;
-   int fKind;
+   std::string fObjectID; ///< unique object identifier
 
 public:
-   TDisplayItem();
-   TDisplayItem(const TDisplayItem &rhs);
-   virtual ~TDisplayItem();
+   TDisplayItem() = default;
+   virtual ~TDisplayItem() {}
 
    void SetObjectIDAsPtr(void *ptr);
-   void SetObjectID(const char *id) { fObjectID = id; }
-   const char *GetObjectID() const { return fObjectID.c_str(); }
-
-   void SetOption(const char *opt) { fOption = opt; }
-   const char *GetOption() { return fOption.c_str(); }
-
-   void SetKind(int kind) { fKind = kind; }
-   int GetKind() const { return fKind; }
+   void SetObjectID(const std::string &id) { fObjectID = id; }
+   std::string GetObjectID() const { return fObjectID; }
 
    static std::string MakeIDFromPtr(void *ptr);
 };
 
-// list of snapshot for primitives in pad
-
-class TPadDisplayItem : public TDisplayItem {
-protected:
-   const TFrame *fFrame{nullptr};         ///! temporary pointer on frame object
-   std::vector<TDisplayItem *> fPrimitives;
-public:
-   TPadDisplayItem() : TDisplayItem(), fPrimitives() { SetKind(3); }
-   virtual ~TPadDisplayItem();
-   void SetFrame(const TFrame *f) { fFrame = f; }
-   void Add(TDisplayItem *snap) { fPrimitives.push_back(snap); }
-   TDisplayItem *Last() const { return fPrimitives[fPrimitives.size() - 1]; }
-   void Clear();
-};
-
 // direct pointer to some object without ownership
 
 template <class T>
 class TOrdinaryDisplayItem : public TDisplayItem {
 protected:
-   const T *fSnapshot;
+   const T *fObject{nullptr}; ///<  direct pointer without ownership
```

Comment:
Are we guaranteed that the lifetime of the `TOrdinaryDisplayItem` exceeds the lifetime of the object `fObject` points to? Or should this be a `weak_ptr` or even `shared_ptr`?

Reply:
> Are we guaranteed that the lifetime of the TOrdinaryDisplayItem exceeds the lifetime of the object fObject points to?

``TOrdinaryDisplayItem`` should be used only for the objects, which exists longer than painter method of canvas/pad are executed. This is typically drawables (or objects inside drawables).
After painting is finished, all display items should be destroyed. 
If we want to use ``weak_ptr`` or ``shared_ptr``, we could create appropriate template classes.
But first we should solve I/O limitation with ``shared_ptr``, while ``TDisplayItem`` have to be streamed and send to client

Reply:
But objects inside drawables can be `weak_ptr`. We get a shared_ptr from that - and now how do we keep that `shared_ptr` alive during the lifetime of the display item? We must do that, because someone else might remove the object, expiring the `weak_ptr` - and how does the display item learn about that?

Reply:
Is weak_ptr supported by ROOT I/O?
Can one create weak_ptr from **this**, without using shared_ptr<TDrawable> which is stored in the list of pad primitives?

Reply:
As I said - we need to wait for I/O support before we can fix this if it needs fixing. And yes, there's `std::enable_shared_from_this`. Unless `this` is `T` - then we need to make `T` a `weak_ptr<T>` (i.e. pass the `weak_ptr` directly to the display item).

We should take note and fix it later.